### DISCUSSION TITLE
Querying multiple attributes of an object in a document

### DISCUSSION THREAD

**Discussion by technocorvus** [ravendb/ravendb#16132] [Q&A]
Good morning.

Wanted to check my understanding of the best (only?) way to query multiple attributes of an object contained in a document.

We have a “Rooms” collection of documents, which among other attributes, contains an array that lists members and their roles:

```
Name: “Test Room”,
Members: [
    {UserId: “User/1”, Role: “Manager”},
    {UserId: “User/2”, Role: “Agent”}
]
```

The query to list all rooms that a given user belongs to is simple enough:

```
from "Rooms" 
where Members[].UserId == "User/1"
```

But _(unless I’m missing something!)_, querying on multiple attributes of an object nested in an array isn’t possible without the creation of a static index; this query _(understandably!)_ returns a result, though it’s not what we’re looking for:

```
from "Rooms" 
where Members[].UserId == "User/1" and Members[].Role == "Agent"
```

With mongo, we’d reach for the `$elemMatch` operator, but there appears to be nothing similar here.

Is the creation of a static index our only option, or am I missing something that will be blindingly obvious in hindsight?

Thanks!

**Comment by ayende** [comment]
There are two options here:
You can do a fanout during the index, or use `intersect`

https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/intersection

```
from Rooms
where intersect(Members[].UserId = "Users/1", Members[].Role = "Agent")
```

**Reply by technocorvus** [reply]
@ayende thanks for the quick response!

We looked at `intersect`, but it doesn’t appear to do what we want.

Given the document:

```
Name: “Test Room”,
Members: [
    {UserId: “User/1”, Role: “Manager”},
    {UserId: “User/2”, Role: “Agent”}
]
```

And query you proposed:

```
from Rooms
where intersect(Members[].UserId = "Users/1", Members[].Role = "Agent”)
```

The result set we’re looking for should be empty (because “User/1” has the role of “Manager”, not “Agent”). However, that query still returns a result.

Is a static index our only option here?

Thanks!

**Reply by garayx** [reply]
Hi, it seems like a bug, I have opened an issue to check that: https://issues.hibernatingrhinos.com/issue/RavenDB-20232

I guess the fanout index is a only option until the issues is resolved.

https://ravendb.net/docs/article-page/5.1/csharp/indexes/fanout-indexes#indexes-fanout-indexes

**Reply by technocorvus** [reply]
@garayx thanks for looking into this.

Based on the example given in https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/intersection, we initially thought `intersect` was our ticket; upon closer reading, a fanout index is created to satisfy the query.

Our current understanding of `intersect` is that it’s performing properly; in my opening example, the document _does_ contain the term `Users/1` in the field `Members[].UserId`, and the term `Agent` in the field `Members[].Role`. That appears to satisfy the condition “…return only documents that match all the provided sub-queries”.

Unless I’m missing something, it feels like a separate/new operator (akin to `$elemMatch`) is necessary to satisfy this type of query in an unambiguous manner.

(The “unless I’m missing something” part is why I’m here; you both know far more about the internals and expected behaviors of the DB!)

**Reply by ayende** [reply]
The easier way to handle that is to do this in an index:

```
from r in docs.Rooms
select new { 
   UserIdAndRole= r.Members.Select(m => new { m.UserId, m.Role})
}
```

And then in your query, do:

```
session.Query<RoomIndex.Result, RoomIndex>().Where(x=>x.UserIdAndRole == new { UserId = uid, Role = role})
  .OfType<Room>()
  .ToList()
```

**Reply by technocorvus** [reply]
@ayende thanks!

Just to check my own understanding: as of 5.4, that’s the only way to query on multiple attributes of an object in a document?

_(Not a complaint, just making sure I understand what’s going on underneath the hood)_

**Reply by ayende** [reply]
Yes, what this actually does is create a compound value, which you can then query directly

**Reply by technocorvus** [reply]
@ayende thanks for the tip; I had no idea such a thing (indexing compound values) was possible!

Is there any additional documentation around those, particularly in regards to index creation/querying via javascript? 

(A search of the [ravendb-nodejs-client](https://github.com/ravendb/ravendb-nodejs-client) repo turned up empty for me, and I couldn’t find any references in the book or the online documentation)

**Reply by ayende** [reply]
You can just return an json object directly, that should work.
Should be something like:

```
 class UsersByNameAndRole extends AbstractJavaScriptIndexCreationTask<User> {
            public constructor() {
                super();
                this.map(User, u => ({ nameAndRole: { u.name, u.role} }));
            }
        }
```

**Reply by ivanlegin** [reply]
> The easier way to handle that is to do this in an index:
> 
> ```
> from r in docs.Rooms
> select new { 
>    UserIdAndRole= r.Members.Select(m => new { m.UserId, m.Role})
> }
> ```
> 
> And then in your query, do:
> 
> ```
> session.Query<RoomIndex.Result, RoomIndex>().Where(x=>x.UserIdAndRole == new { UserId = uid, Role = role})
>   .OfType<Room>()
>   .ToList()
> ```

```
This works and returns expected

        var jsonValue= JsonSerializer.Serialize(new { Season = targetSeason, SelectedCropId = targetCropId });
        
        var query = session.Query<CropRotationsIndex.IndexEntry, CropRotationsIndex>()
        .Where(f => f.SeasonSelectedCropId == jsonValue)
        .OfType<CropRotations>()
        .ToList();

but this no return results:

        var query = session.Query<CropRotationsIndex.IndexEntry, CropRotationsIndex>()
        // I use public object? SeasonSelectedCropId for this case in index, string can not compare with anonymous 
        .Where(f => f.SeasonSelectedCropId == new { Season = targetSeason, SelectedCropId = targetCropId })
        .OfType<CropRotations>()
        .ToList();

why query dosen't serialize Anonymous object in where clause like you presented above
index like: 

public class CropRotationsIndex : AbstractIndexCreationTask<CropRotations, CropRotationsIndex.IndexEntry>
{
public class CropRotationsIndex.IndexEntry
{
    public string? SeasonSelectedCropId { get; set; }
    //public object? SeasonSelectedCropId { get; set; } for annonymous comparison
}
public CropRotationsIndex()
{
this.Map = from field in allFields
                  let cropRotations = LoadDocument<CropRotation>(field.CropRotationIds)
                  select new 
                  {
                     SeasonSelectedCropId = cropRotations.Select(cr => new 
                                                                  { 
                                                                      cr.Season,
                                                                      SelectedCropId = cr.SelectedCropId 
                                                                  })
                  }
}
```

**Reply by garayx** [reply]
@ivanlegin please send a failing unit-test so I can understand what exactly you trying to do,

https://ravendb.net/docs/article-page/5.4/Csharp/start/test-driver

**Reply by ivanlegin** [reply]
It's simple, if i serialize Anonymous object and query this returns results
 
```
 var jsonValue = JsonSerializer.Serialize(new { filter.SelectedCropId, filter.SelectedVarietyId });

        var results = await session.Query<IndexEntry, FieldsByCropRotationsDynamicFieldsIndex>()
        .Where(s => s.SelectedCropVarieties[season] == jsonValue)
        .ToListAsync();
```

same query filter by direct Anonymous object dosen't

        var results = await session.Query<IndexEntry, FieldsByCropRotationsDynamicFieldsIndex>()
        .Where(s => s.SelectedCropVarieties[season] == new { filter.SelectedCropId, filter.SelectedVarietyId })
        .ToListAsync();`

Why this serilization isn't automatic ?

**Reply by garayx** [reply]
It's hard to say what happens from this example. Please send a complete failing test, or simple hello-world application.

https://ravendb.net/docs/article-page/5.4/Csharp/start/test-driver#complete-example
